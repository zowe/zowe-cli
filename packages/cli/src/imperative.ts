/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as path from "path";

import { IImperativeConfig } from "@zowe/imperative";
import { Constants } from "./Constants";

import { ZosmfProfile } from "@zowe/zosmf-for-zowe-sdk";
import { ZosTsoProfile } from "@zowe/zos-tso-for-zowe-sdk";
import { ZosUssProfile } from "@zowe/zos-uss-for-zowe-sdk";
import { ProfileConstants } from "@zowe/core-for-zowe-sdk";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: [
        "daemon/*.definition!(.d).*s",
        "provisioning/*.definition!(.d).*s",
        "workflows/*.definition!(.d).*s",
        "zosconsole/*.definition!(.d).*s",
        "zosfiles/*.definition!(.d).*s",
        "zosjobs/*.definition!(.d).*s",
        "zosmf/*.definition!(.d).*s",
        "zostso/*.definition!(.d).*s",
        "zosuss/*.definition!(.d).*s",
        "zoslogs/*.definition!(.d).*s"
    ],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    webHelpLogoImgPath: path.join(__dirname, "..", "web-help-logo.png"),
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        }
    },
    apimlConnLookup: [
        {
            apiId: "ibm.zosmf",
            gatewayUrl: "api/v1",
            connProfType: "zosmf"
        }
    ],
    baseProfile: {
        ...ProfileConstants.BaseProfile,
        authConfig: [
            {
                serviceName: "apiml",
                handler: __dirname + "/auth/ApimlAuthHandler",
                login: {
                    summary: ProfileConstants.APIML_LOGIN_SUMMARY,
                    description: ProfileConstants.APIML_LOGIN_DESCRIPTION,
                    examples: [
                        ProfileConstants.APIML_LOGIN_EXAMPLE1,
                        ProfileConstants.APIML_LOGIN_EXAMPLE2
                    ],
                    options: [
                        ProfileConstants.BASE_OPTION_HOST,
                        ProfileConstants.BASE_OPTION_PORT,
                        ProfileConstants.BASE_OPTION_USER,
                        ProfileConstants.BASE_OPTION_PASSWORD,
                        ProfileConstants.BASE_OPTION_REJECT_UNAUTHORIZED,
                        ProfileConstants.BASE_OPTION_CERT_FILE,
                        ProfileConstants.BASE_OPTION_CERT_KEY_FILE,
                        ProfileConstants.BASE_OPTION_CERT_ACCOUNT,
                        ProfileConstants.BASE_OPTION_CERT_KEY_ACCOUNT
                    ]
                },
                logout: {
                    summary: ProfileConstants.APIML_LOGOUT_SUMMARY,
                    description: ProfileConstants.APIML_LOGOUT_DESCRIPTION,
                    examples: [
                        ProfileConstants.APIML_LOGOUT_EXAMPLE1,
                        ProfileConstants.APIML_LOGOUT_EXAMPLE2
                    ],
                    options: [
                        ProfileConstants.BASE_OPTION_HOST,
                        ProfileConstants.BASE_OPTION_PORT,
                        ProfileConstants.APIML_LOGOUT_OPTION_TOKEN_TYPE,
                        ProfileConstants.BASE_OPTION_TOKEN_VALUE,
                        ProfileConstants.BASE_OPTION_REJECT_UNAUTHORIZED
                    ]
                }
            }
        ]
    },
    authGroupConfig: {
        authGroup: {
            summary: ProfileConstants.AUTH_GROUP_SUMMARY,
            description: ProfileConstants.AUTH_GROUP_DESCRIPTION
        }
    },
    configAutoInitCommandConfig: {
        handler: __dirname + "/config/auto-init/ApimlAutoInitHandler",
        provider: "APIML",
        autoInit: {
            options: [
                ProfileConstants.AUTO_INIT_OPTION_HOST,
                ProfileConstants.AUTO_INIT_OPTION_PORT,
                ProfileConstants.AUTO_INIT_OPTION_USER,
                ProfileConstants.AUTO_INIT_OPTION_PASSWORD,
                ProfileConstants.AUTO_INIT_OPTION_REJECT_UNAUTHORIZED,
                ProfileConstants.AUTO_INIT_OPTION_TOKEN_TYPE,
                ProfileConstants.AUTO_INIT_OPTION_TOKEN_VALUE,
                ProfileConstants.AUTO_INIT_OPTION_CERT_FILE,
                ProfileConstants.AUTO_INIT_OPTION_CERT_KEY_FILE,
                ProfileConstants.BASE_OPTION_CERT_ACCOUNT,
                ProfileConstants.BASE_OPTION_CERT_KEY_ACCOUNT
            ]
        },
        profileType: "base"
    },
    profiles: [
        ZosmfProfile,
        ZosTsoProfile,
        ZosUssProfile
    ]
};
module.exports = config;
