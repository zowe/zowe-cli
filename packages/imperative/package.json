{
  "name": "@zowe/imperative",
  "version": "8.0.0-next.202402271901",
  "description": "framework for building configurable CLIs",
  "author": "Zowe",
  "license": "EPL-2.0",
  "homepage": "https://github.com/zowe/zowe-cli/tree/master/packages/imperative#readme",
  "bugs": {
    "url": "https://github.com/zowe/zowe-cli/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/zowe/zowe-cli.git"
  },
  "keywords": [
    "CLI",
    "framework",
    "zowe"
  ],
  "files": [
    "lib",
    "web-help/dist",
    "web-diff"
  ],
  "publishConfig": {
    "registry": "https://zowe.jfrog.io/zowe/api/npm/npm-local-release/"
  },
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "scripts": {
    "build:packages": "tsc --pretty",
    "build": "npm run build:packages && npm run build:webHelp",
    "lint": "eslint \"packages/**/*.ts\" \"**/__tests__/**/*.ts\"",
    "lint:packages": "eslint \"packages/**/*.ts\" --ignore-pattern \"**/__tests__/**/*.ts\"",
    "lint:tests": "eslint \"**/__tests__/**/*.ts\"",
    "circularDependencyCheck": "madge -c lib",
    "watch": "concurrently \"tsc --pretty --watch\" \"npm run watch:webHelp\"",
    "build:webHelp": "cd web-help && npm run build",
    "watch:webHelp": "cd web-help && npm run watch",
    "typedoc": "typedoc --options ./typedoc.json ./src/",
    "typedocSpecifySrc": "typedoc --options ./typedoc.json",
    "prepack": "node ../../scripts/prepareLicenses.js",
    "clean": "rimraf lib tsconfig.tsbuildinfo"
  },
  "dependencies": {
    "@types/yargs": "^17.0.32",
    "chalk": "^4.1.2",
    "cli-table3": "^0.6.3",
    "comment-json": "~4.2.3",
    "cross-spawn": "^7.0.3",
    "dataobject-parser": "^1.2.25",
    "deepmerge": "^4.3.1",
    "diff": "^5.2.0",
    "diff2html": "3.4.20-usewontache.1.60e7a2e",
    "fast-glob": "^3.3.2",
    "fastest-levenshtein": "^1.0.16",
    "find-up": "^5.0.0",
    "fs-extra": "^11.0.0",
    "jest-diff": "^29.0.0",
    "js-yaml": "^4.1.0",
    "jsonfile": "^6.0.0",
    "jsonschema": "^1.4.1",
    "lodash": "^4.17.21",
    "lodash-deep": "^2.0.0",
    "log4js": "^6.9.1",
    "markdown-it": "^12.3.2",
    "mustache": "^4.2.0",
    "npm-package-arg": "^11.0.1",
    "opener": "^1.5.2",
    "pacote": "^17.0.6",
    "prettyjson": "^1.2.5",
    "progress": "^2.0.3",
    "read": "^3.0.1",
    "semver": "^7.5.2",
    "stack-trace": "^0.0.10",
    "strip-ansi": "^6.0.1",
    "which": "^4.0.0",
    "wrap-ansi": "^7.0.0",
    "yamljs": "^0.3.0",
    "yargs": "^17.7.2"
  },
  "devDependencies": {
    "@types/cross-spawn": "^6.0.6",
    "@types/diff": "^5.0.9",
    "@types/fs-extra": "^11.0.4",
    "@types/glob": "^8.1.0",
    "@types/jsonfile": "^6.1.4",
    "@types/lodash": "^4.14.202",
    "@types/lodash-deep": "^2.0.4",
    "@types/mustache": "^4.2.5",
    "@types/npm-package-arg": "^6.1.4",
    "@types/pacote": "^11.1.8",
    "@types/progress": "^2.0.7",
    "@types/stack-trace": "^0.0.29",
    "@zowe/secrets-for-zowe-sdk": "8.0.0-next.202402271901",
    "concurrently": "^8.0.0",
    "cowsay": "^1.6.0",
    "deep-diff": "^1.0.0",
    "get-function-arguments": "^1.0.0",
    "serve": "^14.2.1",
    "stream-to-string": "^1.2.0",
    "uuid": "^9.0.1",
    "web-help": "file:web-help",
    "yargs-parser": "^21.1.1"
  },
  "engines": {
    "node": ">=16.7.0"
  }
}
