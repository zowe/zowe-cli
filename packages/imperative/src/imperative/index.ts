/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./env/EnvironmentalVariableSettings";
export * from "./doc/IApimlSvcAttrs";
export * from "./doc/IDaemonContext";
export * from "./doc/IImperativeEnvironmentalVariableSetting";
export * from "./doc/IImperativeEnvironmentalVariableSettings";
export * from "./doc/IImperativeConfig";
export * from "./doc/IImperativeLoggingConfig";
export * from "./ConfigurationLoader";
export * from "./ConfigurationValidator";
export * from "./DefinitionTreeResolver";
export * from "./Imperative";
export * from "./LoggingConfigurer";
export * from "./plugins/AbstractPluginLifeCycle";
export * from "./plugins/PluginManagementFacility";
export * from "./auth/doc/IAuthHandlerApi";
export * from "./auth/handlers/AbstractAuthHandler";
export * from "./auth/handlers/BaseAuthHandler";
export * from "./config/cmd/auto-init/handlers/BaseAutoInitHandler";
