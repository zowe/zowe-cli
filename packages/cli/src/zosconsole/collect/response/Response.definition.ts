/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandDefinition } from "@zowe/imperative";
import { Constants } from "../../../Constants";

export const SyncResponseCommandDefinition: ICommandDefinition = {
    name: "sync-responses",
    aliases: ["sr"],
    summary: "Collect outstanding synchronous console response messages",
    description: `The z/OSMF console REST APIs return a "solicited response key" after successfully issuing ` +
    `a synchronous console command that produces solicited responses. You can use the "solicited response key"` +
    `on the "sync-responses" command to collect any additional outstanding solicited responses ` +
    `from the console the command was issued.` +
    `\n\n` +
    `In general, when issuing a z/OS console command, z/OS applications route responses to the originating console. ` +
    `The command response messages are referred to as "solicited command responses" (i.e. direct responses to the command issued). ` +
    `When issuing a z/OS console command using ${Constants.DISPLAY_NAME}, collection of all solicited command responses ` +
    `is attempted by default. However, there is no z/OS mechanism that indicates the total number of response messages that may be produced ` +
    `from a given command. Therefore, the ${Constants.DISPLAY_NAME} console APIs return a "solicited response key" ` +
    `that can be used to "follow-up" and collect any additional solicited command responses.`,
    type: "command",
    handler: __dirname + "/Response.handler",
    profile: {
        optional: ["zosmf"]
    },
    positionals: [
        {
            name: "responsekey",
            type: "string",
            description: `The "solicited response key" provided in response to a previously issued console command. ` +
            `Used by the z/OSMF console API to collect any additional outstanding solicited responses from a ` +
            `previously issued console command.`,
            regex: "^[a-zA-Z0-9]+$",
            required: true
        }
    ],
    options: [
        {
            name: "console-name",
            aliases: ["cn", "c"],
            description: `The name of the z/OS extended MCS console to direct the command. ` +
            `You must have the required authority to access the console specified. ` +
            `You may also specify an arbitrary name, if your installation allows dynamic creation ` +
            `of consoles with arbitrary names.`,
            type: "string",
            allowableValues: {
                values: [/^[a-zA-Z0-9]+$/.source],
                caseSensitive: false
            }
        }
    ],
    examples: [
        {
            description: "Collect any outstanding additional solicited response messages",
            options: "C4866969"
        }
    ]
};
