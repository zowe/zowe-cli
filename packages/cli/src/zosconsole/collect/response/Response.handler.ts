/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IHandlerParameters } from "@zowe/imperative";
import { CollectCommand, ICollectParms, IConsoleResponse } from "@zowe/zos-console-for-zowe-sdk";
import { ZosmfBaseHandler } from "@zowe/zosmf-for-zowe-sdk";

/**
 * Handle to collect a MVS console command response
 * @export
 * @class Handler
 * @implements {ICommandHandler}
 */
export default class Handler extends ZosmfBaseHandler {

    public async processCmd(commandParameters: IHandlerParameters) {

        const collectParms: ICollectParms = {
            commandResponseKey: commandParameters.arguments.responsekey,
            consoleName: commandParameters.arguments["console-name"]
        };

        const response: IConsoleResponse = await CollectCommand.collect(this.mSession, collectParms);

        // Print out the response
        commandParameters.response.console.log(response.commandResponse);

        // Return as an object when using --response-format-json
        commandParameters.response.data.setObj(response);

    }
}
