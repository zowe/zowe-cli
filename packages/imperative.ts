/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as path from "path";

// WIP Imperative version of Brightside
import { IImperativeConfig } from "@zowe/imperative";
import { Constants } from "./Constants";
import { ZosmfSession } from "./zosmf";
import {
    TSO_OPTION_ACCOUNT_PROFILE,
    TSO_OPTION_CHAR_SET,
    TSO_OPTION_CODE_PAGE,
    TSO_OPTION_COLUMNS,
    TSO_OPTION_LOGON_PROCEDURE,
    TSO_OPTION_REGION_SIZE,
    TSO_OPTION_ROWS
} from "./zostso/src/cli/constants/ZosTso.constants";
import { SshSession } from "./zosuss";
import { ZosFilesOptions } from "./zosfiles/src/cli/ZosFiles.options";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: ["**/cli/*.definition!(.d).*s"],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    webHelpLogoImgPath: path.join(__dirname, "..", "web-help", "logo.png"),
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        }
    },
    baseProfile: {
        type: "base",
        schema: {
            type: "object",
            title: "Base Profile",
            description: "Base profile that stores values shared by multiple service profiles",
            properties: {
                host: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_HOST
                },
                port: {
                    type: "number",
                    optionDefinition: Constants.BASE_OPTION_PORT
                },
                user: {
                    type: "string",
                    secure: true,
                    optionDefinition: Constants.BASE_OPTION_USER
                },
                password: {
                    type: "string",
                    secure: true,
                    optionDefinition: Constants.BASE_OPTION_PASSWORD
                },
                rejectUnauthorized: {
                    type: "boolean",
                    optionDefinition: Constants.BASE_OPTION_REJECT_UNAUTHORIZED
                },
                tokenType: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_TOKEN_TYPE
                },
                tokenValue: {
                    type: "string",
                    secure: true,
                    optionDefinition: Constants.BASE_OPTION_TOKEN_VALUE
                }
            },
            required: []
        },
        createProfileExamples: [
            {
                options: "base1 --host example.com --port 443 --user admin --password 123456",
                description: "Create a profile called 'base1' to connect to host example.com and port 443"
            },
            {
                options: "base2 --host example.com --user admin --password 123456 --reject-unauthorized false",
                description: "Create a profile called 'base2' to connect to host example.com (default port - 443) " +
                    "and allow self-signed certificates"
            },
            {
                options: "base3 --host example.com --port 1443",
                description: "Create a profile called 'base3' to connect to host example.com and port 1443, " +
                    " not specifying a username or password so they are not stored on disk; these will need to be specified on every command"
            },
            {
                options: "base4 --reject-unauthorized false",
                description: "Create a zosmf profile called 'base4' to connect to default port 443 and allow self-signed certificates, " +
                    "not specifying a username, password, or host so they are not stored on disk; these will need to be specified on every command"
            }
        ],
        updateProfileExamples: [
            {
                options: "base1 --user newuser --password newp4ss",
                description: "Update a base profile named 'base1' with a new username and password"
            }
        ],
        authConfig: [
            {
                serviceName: "apiml",
                handler: __dirname + "/auth/src/cli/ApimlAuthHandler",
                login: {
                    summary: Constants.APIML_LOGIN_SUMMARY,
                    description: Constants.APIML_LOGIN_DESCRIPTION,
                    examples: [
                        Constants.APIML_LOGIN_EXAMPLE1,
                        Constants.APIML_LOGIN_EXAMPLE2
                    ],
                    options: [
                        Constants.BASE_OPTION_HOST,
                        Constants.BASE_OPTION_PORT,
                        Constants.BASE_OPTION_USER,
                        Constants.BASE_OPTION_PASSWORD,
                        Constants.BASE_OPTION_REJECT_UNAUTHORIZED
                    ]
                },
                logout: {
                    summary: Constants.APIML_LOGOUT_SUMMARY,
                    description: Constants.APIML_LOGOUT_DESCRIPTION,
                    examples: [
                        Constants.APIML_LOGOUT_EXAMPLE1,
                        Constants.APIML_LOGOUT_EXAMPLE2
                    ],
                    options: [
                        Constants.BASE_OPTION_HOST,
                        Constants.BASE_OPTION_PORT,
                        Constants.APIML_LOGOUT_OPTION_TOKEN_TYPE,
                        Constants.BASE_OPTION_TOKEN_VALUE,
                        Constants.BASE_OPTION_REJECT_UNAUTHORIZED
                    ]
                }
            }
        ]
    },
    authGroupConfig: {
        authGroup: {
            summary: Constants.AUTH_GROUP_SUMMARY,
            description: Constants.AUTH_GROUP_DESCRIPTION
        }
    },
    profiles: [
        {
            type: "zosmf",
            schema: {
                type: "object",
                title: "z/OSMF Profile",
                description: "z/OSMF Profile",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_HOST_PROFILE
                    },
                    port: {
                        type: "number",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PORT
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_USER_PROFILE
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PASSWORD_PROFILE
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_REJECT_UNAUTHORIZED
                    },
                    basePath: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_BASE_PATH
                    },
                    encoding: {
                        type: "number",
                        optionDefinition: {
                            name: "encoding",
                            aliases: ["ec"],
                            description: "The encoding for download and upload of z/OS data set and USS files." +
                                " The default encoding if not specified is 1047.",
                            type: "number"
                        }
                    },
                    responseTimeout: {
                        type: "number",
                        optionDefinition: ZosFilesOptions.responseTimeout
                    }
                },
                required: []
            },
            createProfileExamples: [
                {
                    options: "zos123 --host zos123 --port 1443 --user ibmuser --password myp4ss",
                    description: "Create a zosmf profile called 'zos123' to connect to z/OSMF at host zos123 and port 1443"
                },
                {
                    options: "zos124 --host zos124 --user ibmuser --password myp4ss --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                },
                {
                    options: "zos125 --host zos125 --port 1443",
                    description: "Create a zosmf profile called 'zos125' to connect to z/OSMF at the host zos125 and port 1443, " +
                        " not specifying a username or password so they are not stored on disk; these will need to be specified on every command"
                },
                {
                    options: "zos126 --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos126' to connect to z/OSMF on the default port 443 and allow self-signed certificates, " +
                        "not specifying a username, password, or host so they are not stored on disk; these will need to be specified on every command"
                },
                {
                    options: "zosAPIML --host zosAPIML --port 2020 --user ibmuser --password myp4ss --reject-unauthorized false --base-path basePath",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                }
            ],
            updateProfileExamples: [
                {
                    options: "zos123 --user newuser --password newp4ss",
                    description: "Update a zosmf profile named 'zos123' with a new username and password"
                }
            ]
        },
        {
            type: "tso",
            schema: {
                type: "object",
                title: "TSO Profile",
                description: "z/OS TSO/E User Profile",
                properties: {
                    account: {
                        type: "string",
                        optionDefinition: TSO_OPTION_ACCOUNT_PROFILE
                    },
                    characterSet: {
                        type: "string",
                        optionDefinition: TSO_OPTION_CHAR_SET
                    },
                    codePage: {
                        type: "string",
                        optionDefinition: TSO_OPTION_CODE_PAGE
                    },
                    columns: {
                        type: "number",
                        optionDefinition: TSO_OPTION_COLUMNS
                    },
                    logonProcedure: {
                        type: "string",
                        optionDefinition: TSO_OPTION_LOGON_PROCEDURE
                    },
                    regionSize: {
                        type: "number",
                        optionDefinition: TSO_OPTION_REGION_SIZE
                    },
                    rows: {
                        type: "number",
                        optionDefinition: TSO_OPTION_ROWS
                    }
                },
                required: []
            },
            createProfileExamples: [
                {
                    description: "Create a tso profile called 'myprof' with default settings and JES accounting information of 'IZUACCT'",
                    options: "myprof -a IZUACCT"
                },
                {
                    description: "Create a tso profile called 'largeregion' with a region size of 8192, a logon procedure of MYPROC, and " +
                        "JES accounting information of '1234'",
                    options: "largeregion -a 1234 --rs 8192"
                },
                {
                    description: "Create a tso profile called 'myprof2' with default settings and region size of 8192, without storing the user account on disk",
                    options: "myprof2 --rs 8192"
                }
            ],
            updateProfileExamples: [
                {
                    description: "Update a tso profile called myprof with new JES accounting information",
                    options: "myprof -a NEWACCT"
                }
            ]
        },
        {
            type: "ssh",
            schema: {
                type: "object",
                title: "z/OS SSH Profile",
                description: "z/OS SSH Profile",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: SshSession.SSH_OPTION_HOST_PROFILE
                    },
                    port: {
                        type: "number",
                        optionDefinition: SshSession.SSH_OPTION_PORT
                    },
                    user: {
                        type: "string",
                        optionDefinition: SshSession.SSH_OPTION_USER_PROFILE
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: SshSession.SSH_OPTION_PASSWORD
                    },
                    privateKey: {
                        type: "string",
                        optionDefinition: SshSession.SSH_OPTION_PRIVATEKEY
                    },
                    keyPassphrase: {
                        type: "string",
                        secure: true,
                        optionDefinition: SshSession.SSH_OPTION_KEYPASSPHRASE
                    },
                    handshakeTimeout: {
                        type: "number",
                        optionDefinition: SshSession.SSH_OPTION_HANDSHAKETIMEOUT
                    }
                },
                required: []
            },
            createProfileExamples: [
                {
                    options: "ssh111 --host sshhost --user ibmuser --password myp4ss",
                    description: "Create a ssh profile called 'ssh111' to connect to z/OS SSH server at host 'zos123' and default port 22"
                },
                {
                    options: "ssh222 --host sshhost --port 13022 --user ibmuser --password myp4ss",
                    description: "Create a ssh profile called 'ssh222' to connect to z/OS SSH server at host 'zos123' and port 13022"
                },
                {
                    options: "ssh333 --host sshhost --user ibmuser --privateKey /path/to/privatekey --keyPassphrase privateKeyPassphrase",
                    description: "Create a ssh profile called 'ssh333' to connect to z/OS SSH server at host 'zos123' " +
                        "using a privatekey '/path/to/privatekey' and its decryption passphrase 'privateKeyPassphrase' " +
                        "for privatekey authentication"
                },
                {
                    options: "ssh444 --privateKey /path/to/privatekey",
                    description: "Create a ssh profile called 'ssh444' to connect to z/OS SSH server on default port 22, without specifying " +
                        "username, host, or password, preventing those values from being stored on disk"
                }
            ]
        }

    ]
};
module.exports = config;
