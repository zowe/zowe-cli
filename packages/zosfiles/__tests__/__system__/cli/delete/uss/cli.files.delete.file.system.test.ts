/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { Imperative, Session } from "@zowe/imperative";
import { Create } from "../../../../..";
import { inspect } from "util";
import { runCliScript } from "../../../../../../../__tests__/__src__/TestUtils";
import { TestEnvironment } from "../../../../../../../__tests__/__src__/environment/TestEnvironment";
import { ITestEnvironment } from "../../../../../../../__tests__/__src__/environment/doc/response/ITestEnvironment";
import { ITestPropertiesSchema } from "../../../../../../../__tests__/__src__/properties/ITestPropertiesSchema";

let REAL_SESSION: Session;
// Test Environment populated in the beforeAll();
let TEST_ENVIRONMENT: ITestEnvironment;
let TEST_ENVIRONMENT_NO_PROF: ITestEnvironment;
let defaultSystem: ITestPropertiesSchema;
let ussname: string;
let user: string;

describe("Delete File", () => {

    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            tempProfileTypes: ["zosmf"],
            testName: "zos_delete_uss_file"
        });

        defaultSystem = TEST_ENVIRONMENT.systemTestProperties;

        REAL_SESSION = TestEnvironment.createZosmfSession(TEST_ENVIRONMENT);
        user = defaultSystem.zosmf.user.trim();
        ussname = `${defaultSystem.zosmf.user.trim()}.aTestUssFileDelete.txt`;
        ussname = `${defaultSystem.unix.testdir}/${ussname}`;
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
    });

    describe("Success scenario without profiles", () => {
        let defaultSys: ITestPropertiesSchema;

        // Create the unique test environment
        beforeEach(async () => {
            TEST_ENVIRONMENT_NO_PROF = await TestEnvironment.setUp({
                testName: "zos_files_delete_uss_file_without_profile"
            });

            defaultSys = TEST_ENVIRONMENT_NO_PROF.systemTestProperties;
            let error;
            let response;
            try {
                response = await Create.uss(REAL_SESSION, ussname, "file");
            } catch (err) {
                error = err;
                Imperative.console.info("Error: " + inspect(error));
            }
        });

        afterAll(async () => {
            await TestEnvironment.cleanUp(TEST_ENVIRONMENT_NO_PROF);
        });

        it("should delete a uss file, with fully specified command", async () => {
            const ZOWE_OPT_BASE_PATH = "ZOWE_OPT_BASE_PATH";

            // if API Mediation layer is being used (basePath has a value) then
            // set an ENVIRONMENT variable to be used by zowe.
            if (defaultSys.zosmf.basePath != null) {
                TEST_ENVIRONMENT_NO_PROF.env[ZOWE_OPT_BASE_PATH] = defaultSys.zosmf.basePath;
            }

            const response = runCliScript(__dirname + "/__scripts__/command/command_delete_file_fully_qualified.sh",
                TEST_ENVIRONMENT_NO_PROF, [ussname, "--for-sure",
                    defaultSys.zosmf.host,
                    defaultSys.zosmf.port,
                    defaultSys.zosmf.user,
                    defaultSys.zosmf.pass]);
            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout.toString()).toMatchSnapshot();
        });
    });

    describe("Success scenarios", () => {
        let defaultSys: ITestPropertiesSchema;

        // Create the unique test environment
        beforeEach(async () => {
            TEST_ENVIRONMENT_NO_PROF = await TestEnvironment.setUp({
                testName: "zos_files_delete_uss_file_without_profile"
            });

            defaultSys = TEST_ENVIRONMENT_NO_PROF.systemTestProperties;
            let error;
            let response;

            try {
                response = await Create.uss(REAL_SESSION, ussname, "file");
            } catch (err) {
                error = err;
                Imperative.console.info("Error: " + inspect(error));
            }
        });

        afterAll(async () => {
            await TestEnvironment.cleanUp(TEST_ENVIRONMENT_NO_PROF);
        });

        it("should delete a file", async () => {
            const response = runCliScript(__dirname + "/__scripts__/command/command_delete_file.sh",
                TEST_ENVIRONMENT, [ussname, "--for-sure"]);
            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout.toString()).toMatchSnapshot();
        });
    });

    describe("Expected failures", () => {
        it("should fail deleting a file that does not exist", async () => {
            const response = runCliScript(__dirname + "/__scripts__/command/command_delete_file.sh",
                TEST_ENVIRONMENT, [ussname, "--for-sure"]);
            expect(response.status).toBe(1);
            expect(response.stderr.toString()).toContain("No such file or directory.");
            expect(response.stdout.toString()).toMatchSnapshot();
        });
    });
});
