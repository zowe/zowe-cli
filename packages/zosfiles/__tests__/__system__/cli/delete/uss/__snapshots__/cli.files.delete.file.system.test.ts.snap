// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete File Expected failures should fail deleting a file that does not exist 1`] = `
"================Z/OS FILES DELETE FILE===============
"
`;

exports[`Delete File Success scenario without profiles should delete a uss file, with fully specified command 1`] = `
"================Z/OS FILES DELETE FILE FULLY SPECIFIED===============
USS File or directory deleted successfully.
"
`;

exports[`Delete File Success scenarios should delete a file 1`] = `
"================Z/OS FILES DELETE FILE===============
USS File or directory deleted successfully.
"
`;

exports[`Delete File Success scenarios should delete a file with response timeout 1`] = `
"================Z/OS FILES DELETE FILE===============
USS File or directory deleted successfully.
"
`;
